### PR TITLE
Lint for significant drops who may have surprising lifetimes

### PR DESCRIPTION
**Pull request by PrestonFrom**
Add new trait (`SignificantDrop`) and lint to look for temporaries of types who implement it and who may have surprising lifetimes.

```
changelog: new lint [`significant_drop_in_scrutinee`] 
```

### CODE REVIEW

File: library/core/src/marker.rs [hunk 1]
```diff
@@ -830,3 +830,9 @@ mod copy_impls {
     #[stable(feature = "rust1", since = "1.0.0")]
     impl<T: ?Sized> Copy for &T {}
 }
+
+/// Trait to help linters look for drops that happen at unexpected times
+/// Any type whose drop has a significant side effect can impl this trait
+/// so linters can let users know if it has a potentially surprising lifetime.
+#[stable(feature = "significant_drop", since = "1.60.0")]
```
Comment (oli-obk) [inline library/core/src/marker.rs]: Please make it unstable for now

File: src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                has_significant_drop(&exprs[0], did, context)
+            }
+            _ => {
+                false
+            }
+        }
+    } else {
+        found_significant_drop
+    }
+}
```
Comment (oli-obk) [inline src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs]: You probably need to format these files, but I'm unsure how to do that in the clippy copy inside the rustc repo

File: src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+/// implements `SignificantDrop`. If one is found, return true. Otherwise, return false.
+fn has_significant_drop<'a, 'tcx>(expression: &Expr<'a>, did: DefId, context: &LateContext<'tcx>) -> bool{
+    let found_significant_drop = {
+        let self_ty = context.typeck_results().expr_ty(expression);
+
+        implements_trait(context, self_ty, did, &[])
+
+    };
+    if !found_significant_drop {
+        match expression.peel_drop_temps().kind {
```
Comment (oli-obk) [inline src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs]: Whoa, magic function, good find!

File: src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub SIGNIFICANT_DROP_IN_SCRUTINEE,
+    nursery,
+    "default lint description"
```
Comment (oli-obk) [inline src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs]: You forgot to fill in this placeholder

**Review by oli-obk — CHANGES_REQUESTED** [review]
While I have a few more things that we shoukd iterate on, (message text, more precise spans, documentation) I think we should merge it and then keep working on it in follow ups. This should put less of a burden on you and give everyone something to experiment with, even if it's not yet where we want it to be at the end

File: library/core/src/marker.rs (hunk 1 above)
Comment (PrestonFrom) [reply library/core/src/marker.rs]: Done!

File: src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs (hunk 2 above)
Comment (PrestonFrom) [reply src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs]: I checked the clippy repo documentation, but I couldn't find anything on formatting, so I ran `cargo fmt`. I'm not sure if that will be satisfactory or not though.

File: src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs (hunk 3 above)
Comment (PrestonFrom) [reply src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs]: :)

File: src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs (hunk 4 above)
Comment (PrestonFrom) [reply src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs]: Oh! Thank you!

**Comment by PrestonFrom** [comment]
Thank you for the initial review -- reviewing/merging an unstable version and iterating makes sense to me. I'll open a PR into the main repo.

**Pull request closed without merging**